src/pubsub/ua_pubsub_keystorage.c

```diff
@@ -0,0 +1,530 @@
+                                      UA_UInt32 currentTokenId, const UA_ByteString *currentKey,
+                                      UA_ByteString *futureKeys, size_t futureKeyCount,
+                                      UA_Duration msKeyLifeTime) {
+    UA_StatusCode retval = UA_STATUSCODE_BAD;
+
+    if(!server || !securityGroupId) {
+        retval = UA_STATUSCODE_BADINVALIDARGUMENT;
+        goto error;
+    }
+
+    if(futureKeyCount > 0 && !futureKeys) {
+        retval = UA_STATUSCODE_BADARGUMENTSMISSING;
+        goto error;
+    }
+
+    size_t keyNumber = futureKeyCount;
+
+    UA_PubSubKeyStorage *keyStorage =
+        UA_PubSubKeyStorage_getKeyStorageWithSecurityGroupId(server, *securityGroupId);
+    if(!keyStorage) {
+        retval = UA_STATUSCODE_BADNOTFOUND;
+        goto error;
+    }
+
+    if(currentKey && keyStorage->keyListSize == 0) {
+        keyStorage->firstItem =
+            (UA_PubSubKeyListItem *)UA_malloc(sizeof(UA_PubSubKeyListItem));
+        if(!keyStorage->firstItem)
+            goto error;
+
+        keyStorage->firstItem->next = NULL;
+
+        keyStorage->firstItem->keyID = currentTokenId;
+        retval = UA_ByteString_copy(currentKey, &keyStorage->firstItem->key);
+        if(UA_StatusCode_isBad(retval))
+            goto error;
+
+        keyStorage->keyListSize++;
+    }
+
+    UA_PubSubKeyListItem *keyListIterator = keyStorage->firstItem;
+    UA_UInt32 startingTokenID = currentTokenId + 1;
+    for(size_t i = 0; i < keyNumber; ++i) {
+        retval = UA_PubSubKeyStorage_getKeyByKeyID(
+            startingTokenID, keyStorage->firstItem, &keyListIterator);
+        /*Skipping key with matching KeyID in existing list*/
+        if(retval == UA_STATUSCODE_BADNOTFOUND) {
+            keyListIterator = UA_PubSubKeyStorage_push(keyListIterator, &futureKeys[i], startingTokenID);
+            if(!keyListIterator)
+                goto error;
+
+            keyStorage->lastItem = keyListIterator;
+            keyStorage->keyListSize++;
+        }
+        if(startingTokenID == UA_UINT32_MAX)
+            startingTokenID = 1;
+        else
+            ++startingTokenID;
+    }
+
+    /*update keystorage references*/
+    retval = UA_PubSubKeyStorage_getKeyByKeyID(currentTokenId, keyStorage->firstItem, &keyStorage->currentItem);
+    if (retval != UA_STATUSCODE_GOOD && !keyStorage->currentItem)
+        goto error;
+
+    UA_PubSubKeyListItem *lastItem = keyStorage->firstItem;
+    while (lastItem->next != NULL) {
+        lastItem = lastItem->next;
+    }
+
+    keyStorage->lastItem = lastItem;
+    keyStorage->keyList = keyStorage->firstItem;
+
+    keyStorage->keyLifeTime = msKeyLifeTime;
+
+    return retval;
+error:
+    if(keyStorage) {
+        UA_PubSubKeyStorage_clearKeyList(keyStorage);
+    }
+    return retval;
+}
+
+UA_StatusCode
+UA_PubSubKeyStorage_getKeyByKeyID(const UA_UInt32 keyId, UA_PubSubKeyListItem *firstItem,
```

Comment:
In the sense of a "quasi object-oriented" API, forward a UA_PubSubKeyStorage pointer to this method.
Instead of forwarding only elements from a UA_PubSubKeyStorage.
That also allows us the change the internal definitions of UA_PubSubKeyStorage without changing this API.

This applies to all methods with the UA_PubSubKeyStorage_ prefix.

Reply:
Whereever possible changed to passing PubSubKeyStorage pointer.